src/kadmin/cli/keytab.c

```diff
@@ -277,11 +266,23 @@ add_principal(void *lhandle, char *keytab_str, krb5_keytab keytab,
         goto cleanup;
     }
 
-#ifdef KADMIN_LOCAL
-    if (norandkey)
-        code = kadm5_get_principal_keys(handle, princ, &keys, &nkeys);
-    else
-#endif
+    if (norandkey) {
+        kadm5_key_data *key_data;
+
+        code = kadm5_get_principal_keys(handle, princ, 0, &key_data, &nkeys);
+        if (code == 0) {
+            keys = malloc(nkeys * sizeof(krb5_keyblock));
+            if (!keys) {
+                com_err(whoami, ENOMEM, _("while extracting keys"));
+                goto cleanup;
+            }
+            for (i = 0; i < nkeys; i++) {
+                keys[i] = key_data[i].key;
+                memset(&key_data[i].key, 0 , sizeof(krb5_keyblock));
```

Comment:
Because it's easier to see that there's no mistake here.  At first I thought `key_data` might not be a krb5_keyblock, and I had to go check.  Using a `sizeof()` expression in a `memset()` call such that the expression in the `sizeof()` is derived from the first argument to `memset()` is a time-honored idiom for avoiding using the wrong size.  You might think "who cares, I got it right in this case", but other people will have to read this code later (making it easier to read the code is a nice courtesy to extend to them), and some might have to change it, and they might screw it up.